source/classes.tex

```diff
@@ -6681,12 +6707,72 @@
 derived-to-base conversions\iref{over.best.ics},
 class member access expressions\iref{expr.ref}, and
 array subscript expressions\iref{expr.sub} applied to \tcode{x}.
-The type of the expression $\tcode{x}_i$ \tcode{<=>} $\tcode{x}_i$
-is denoted by $\tcode{R}_i$.
-It is unspecified
-whether virtual base class subobjects are compared more than once.
+It is unspecified whether virtual base class subobjects
+appear more than once in the expanded list of subobjects.
+
+\rSec2[class.eq]{Equality operators}
+\indextext{operator!equality!defaulted}%
+\indextext{operator!inequality!defaulted}%
+
+\pnum
+A defaulted equality operator\iref{expr.eq} function
+shall have a declared return type \tcode{bool}.
+
+\pnum
+A defaulted \tcode{==} operator function for a class \tcode{C}
+is defined as deleted
+unless, for each $\tcode{x}_i$ in the expanded list of subobjects
+for an object \tcode{x} of type \tcode{C},
+$\tcode{x}_i\tcode{ == }\tcode{x}_i$ is a valid expression and
+contextually convertible to \tcode{bool}.
 
 \pnum
+The return value \tcode{V} of a defaulted \tcode{==} operator function
+with parameters \tcode{x} and \tcode{y} is determined
+by comparing corresponding elements $\tcode{x}_i$ and $\tcode{y}_i$
+in the expanded lists of subobjects for \tcode{x} and \tcode{y}
```

Comment:
This notably doesn't specify the order in which corresponding elements are compared. Without an ordering, this wording allows comparing the same pair of corresponding elements more than once. @BRevzin, is it intended that the comparison order is unspecified, or should we have something like "in order by increasing subscript" here?

Reply:
Is that not implied by "until the first index `i` where ..."? 

If not, we need to add "in order by increasing subscript" to [class.spaceship]/3 as well, since there's where I copied this wording from.

Reply:
I don't think we should address this (if it needs addressing at all) in the paper merge cycle.

Reply:
> Is that not implied by "until the first index `i` where ..."?

That tells me that the process terminates upon finding such an index, it doesn't specify the order in which indices are chosen.

> If not, we need to add "in order by increasing subscript" to [class.spaceship]/3 as well, since there's where I copied this wording from.

+1

> I don't think we should address this (if it needs addressing at all) in the paper merge cycle.

I'd be happy to merge this as is and follow up with a core issue to determine the ordering if you don't think this falls under "editorial clarification of the wording".

Reply:
Please create a separate editorial issue for this. Let's see whether Richard is willing to address this editorially.

Reply:
Filed as https://github.com/cplusplus/draft/issues/2748. Feel free to mark this discussion "resolved".